### DISCUSSION TITLE
setValue not working on controlled form

### DISCUSSION THREAD

**Discussion by gilneto8** [react-hook-form/react-hook-form#6797] [Q&A] [answered]
Hi everyone.

I'll try to explain this simply, but since this is starting to get rather complex, bear with me a bit.

**TLDR**: I have a controlled form and I can't seem to make `setValue` programmatically update the value of a registered field. When I try to invoke `handleSubmit`, it remains with value `undefined`.

----

**Not TLDR**:

All fields in this form have type `LinkedField`, which is set as `type LinkedField<T> = T | string`. This is useful because, on our case, we might want to **link** certain fields to certain environment variables' values and, when that's the case, we just keep a reference to it directly in the field (in the form of `${<variablename>}`). If the field is **linked**, then it's set as disabled until it is manually **unlinked**.

The linking logic is all managed within a component called `ConnectableLabel`, and it's guaranteed that it returns the variable's name to link the field to, and executes the callback (or `action` in this case) when prompted to save the result.

Now for specifics: here, we're talking about a form field called `loggingMessage` which has type `LinkedField<string>`. Redundant, I know. The form itself is implemented as follows (omitting some irrelevant parts):

```typescript

  const [disabled, setDisabled] = useState<{ [key: string]: boolean }>({});

  const resolver = rhfResolver(schema) as unknown as (
    values: Properties,
  ) => ResolverResult<Properties>;

  const form = useForm<Properties>({
    resolver,
    defaultValues: data,
  });

  // ...

  const link = (field: keyof Properties, value: string) => {
    console.log('field value', field, value);
    form.setValue(field, value, { shouldValidate: false });
    setDisabled({ ...disabled, [field]: !!value });
  };

  // ...

  return (
    <Form className={'form p-3 w-100'} onSubmit={form.handleSubmit(action)}>
      <Form.Group>
        <ConnectableLabel
          label={'Logging Message'}
          action={(selection) => link('loggingMessage', selection)}
          value={
            disabled['loggingMessage']
              ? form.getValues('loggingMessage')
              : undefined
          }
          node={node}
        />
        <Form.Control
          disabled={disabled['loggingMessage']}
          as="textarea"
          {...form.register('loggingMessage')}
        />
      </Form.Group>
      <hr className={'hr mt-1 mb-3'} />
      {editable && (
        <Button className={'mt-2'} type={'submit'}>
          Save
        </Button>
      )}
    </Form>
  );

```

I believe the logic here is correct, and most of it works: the field is available for editing as long as it's not disabled, which it becomes if it is linked with an external value, linkage that is set by the function `link()`. So far so good.

The problem appears here: the line `form.setValue(field, value, { shouldValidate: false });` is not actually setting anything in the field (when I do `form.getValues('loggingMessage')`). When I submit the form and it gets caught by the parent component, I do a `console.log` to check the content and `loggingMessage` comes out `undefined`, even after actually confirming that the `link()` function receives both the correct field and value.

![image](https://user-images.githubusercontent.com/62651996/137193588-f7a1e475-0ffd-4852-badd-248c1da66531.png)
![image](https://user-images.githubusercontent.com/62651996/137193615-d59fda44-eb28-47ec-863f-b57508b0e9a4.png)

What is even stranger is that, after saving the variable linkage (and, as such, after executing its callback), the input box is actually updated with the correct variable reference.

![image](https://user-images.githubusercontent.com/62651996/137193508-a0d06106-fc70-41a7-8162-14b4077e260f.png)

So as you might guess, I'm at a loss of ideas. Any input is welcome, and it would be great if you could help me.

Thanks!

**Comment by bluebill1049** [comment, marked as answer]
for the controlled component, you should wrap it with `Controller` or using `useController` hook.

https://react-hook-form.com/api/usecontroller
https://react-hook-form.com/api/usecontroller/controller

**Reply by gilneto8** [reply]
Thanks @bluebill1049, that was the last piece of the puzzle.

For future reference, all I've done was change the code above to be as follows:

```typescript

  // imported the Controller component
  import { useForm, Controller } from 'react-hook-form';

  // ...


  // new argument passed here, the 'onChange' function from the controlled field...
  const link = (field: keyof Properties, value: string, fn: (value: string) => void) => {
    // ... which is executed with the new value
    fn(value);
    setDisabled({ ...disabled, [field]: !!value });
  };

  // ...

  return (
    <Form className={'form p-3 w-100'} onSubmit={form.handleSubmit(action)}>
      <Form.Group>
      	{/* instead of rendering the field directly, rendered it as a child of Controller */}
        <Controller<Properties>
          name={'loggingMessage'}
          control={form.control}
        	{/* rendering the controlled field and not the 'uncontrolled' version*/}
          render={({ field }) => (
            <>
              <ConnectableLabel
                label={'Logging Message'}
                action={(selection) =>
                	// invoked the new version of the link() function
                  link('loggingMessage', selection, field.onChange)
                }
                value={
                  disabled[field.name]
                    ? form.getValues(field.name)
                    : undefined
                }
                node={node}
              />
              <Form.Control
                {...field}
                disabled={disabled[field.name]}
                as="textarea"
              />
            </>
          )}
        />
      </Form.Group>
      <hr className={'hr mt-1 mb-3'} />
      {editable && (
        <Button className={'mt-2'} type={'submit'}>
          Save
        </Button>
      )}
    </Form>
  );

```

I'll abstract some things and iterate on others, but this, as it is, is working. 

Thank you so much!

**Reply by bluebill1049** [reply]
thanks for sharing solution and glad it's working out.
